feat: debian 11 support

jooola opened the pull request:
This is heavily inspired by #1168.

Closes #1271 

Needs more testing, specially on the following topics:
 - [x] playout has troubles ~~with the api version~~ starting. (See #1293)
- [x] ~~I don't use vagrant, so someone should test the vagrant stuff.~~ Vagrant is working


I think #1168 could reuse some of this PR to install the proper php packages.

paddatrapper commented:
>playout has troubles with the api version

That output is expected and doesn't show an error?

I can take a look at the vagrant testing. The issue with Ubuntu 20.04 is a broken liquidsoap version (there is an incorrect symlink). I need to see how to get an update into 20.04 to fix it.

paddatrapper commented on installer/apt/install-debian-php-sury.sh:
```diff
@@ -0,0 +1,36 @@
+  ca-certificates \
+  curl
+
+wget -O "/etc/apt/trusted.gpg.d/php.gpg" "https://packages.sury.org/php/apt.gpg"
```
This should go in /usr/share/keyrings and the apt sources line follow https://wiki.debian.org/DebianRepository/UseThirdParty

jooola replied:
Oh nice to know these recommendation !

Just updated the changes.

paddatrapper commented on installer/apt/install-debian-php-sury.sh:
```diff
@@ -0,0 +1,36 @@
+
+apt-get update
+
+# Install packages
```
Why are you breaking this out of the standard apt package list?

jooola replied:
Because we need extra care not to install PHP versions such as 7.4 or 8.0, as the management of those php packages is tricky.

See #1256 

You can also find some guidance: From this anchor down https://github.com/oerdnj/deb.sury.org/wiki/Frequently-Asked-Questions#php-pear

jooola replied:
I think this is the same for #1168, we should check it doesn't mess up everything.

paddatrapper replied:
Yeah, but we can do that by specifying the versions required in the apt list can we not?

jooola replied:
I haven't found any other way to prevent installing php8.0 and php7.4 when installing php-pear or php-apcu.

But feel free to tweak the this part.

jooola replied:
I'll try again to find some better way some other time. But I have to admit this small hacky script blend perfectly in with the current install script mess :laughing:

jooola replied:
- `php-pear`: adding php7.3-xml is preventing any other php version to be installed. Maybe this could be added to the list.
- `php7.3-apcu`: needs to be installed with the `--no-install-recommends` flags otherwise the default php cli will be installed (7.4 or 8.0).

paddatrapper replied:
I see. Then all except `php7.3-apcu` can be added to the standard list, as they all get installed as a single `apt-get install` call, or am I still not understanding?

jooola replied:
@paddatrapper Yes, no misunderstanding.

I prefer not to mix this logic to the huge apt package list. This feel clean to me, PHP stuff in a separate script.
But if you think we should merge everything back in the install script I can do that.

paddatrapper replied:
I think, until #1246 works and is merged, we should keep things consistent and have as much in the apt-requirements file. Then only have things that require `--no-recommends` elsewhere.

jooola commented:
> That output is expected and doesn't show an error?

No I don't show any error, but haven't looked at it in details. The service never really starts and stops right after this version check. I guess it can be somewhere later in the code. Here is the service, saying it failed to many times :
```sh
jo@debian11:~$ sudo systemctl status libretime-playout.service 
[sudo] password for jo: 
● libretime-playout.service - Libretime Playout Service
     Loaded: loaded (/etc/systemd/system/libretime-playout.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Sun 2021-08-15 17:09:07 CEST; 1h 9min ago
    Process: 3632 ExecStart=/usr/bin/airtime-playout (code=exited, status=1/FAILURE)
   Main PID: 3632 (code=exited, status=1/FAILURE)
        CPU: 358ms

Aug 15 17:09:07 debian11 systemd[1]: libretime-playout.service: Scheduled restart job, restart counter is at 5.
Aug 15 17:09:07 debian11 systemd[1]: Stopped Libretime Playout Service.
Aug 15 17:09:07 debian11 systemd[1]: libretime-playout.service: Start request repeated too quickly.
Aug 15 17:09:07 debian11 systemd[1]: libretime-playout.service: Failed with result 'exit-code'.
Aug 15 17:09:07 debian11 systemd[1]: Failed to start Libretime Playout Service.
```

jooola commented:
Here are some more useful logs:
```
Aug 15 18:41:07 debian11 systemd[1]: Started Libretime Playout Service.
Aug 15 18:41:07 debian11 airtime-playout[11261]: 2021-08-15 18:41:07,818 [__main__] [INFO ]  ###########################################
Aug 15 18:41:07 debian11 airtime-playout[11261]: 2021-08-15 18:41:07,818 [__main__] [INFO ]  #             *** pypo  ***               #
Aug 15 18:41:07 debian11 airtime-playout[11261]: 2021-08-15 18:41:07,818 [__main__] [INFO ]  #   Liquidsoap Scheduled Playout System   #
Aug 15 18:41:07 debian11 airtime-playout[11261]: 2021-08-15 18:41:07,818 [__main__] [INFO ]  ###########################################
Aug 15 18:41:07 debian11 airtime-playout[11261]: 2021-08-15 18:41:07,818 [__main__] [INFO ]  Timezone: ('CET', 'CEST')
Aug 15 18:41:07 debian11 airtime-playout[11261]: 2021-08-15 18:41:07,818 [__main__] [INFO ]  UTC time: 2021-08-15 16:41:07.818735
Aug 15 18:41:07 debian11 airtime-playout[11261]: 2021-08-15 18:41:07,846 [version1] [INFO ]  Airtime API version found: 1.1
Aug 15 18:41:07 debian11 airtime-playout[11261]: 2021-08-15 18:41:07,846 [version1] [INFO ]  pypo is only compatible with API version: 1.1
Aug 15 18:41:07 debian11 airtime-liquidsoap[3315]: 2021/08/15 18:41:07 [server:3] New client: 127.0.0.1.
Aug 15 18:41:07 debian11 airtime-liquidsoap[3315]: 2021/08/15 18:41:07 [server:3] Client 127.0.0.1 disconnected.
Aug 15 18:41:07 debian11 airtime-playout[11261]: Traceback (most recent call last):
Aug 15 18:41:07 debian11 airtime-playout[11261]:   File "/usr/bin/airtime-playout", line 4, in <module>
Aug 15 18:41:07 debian11 airtime-playout[11261]:     __import__('pkg_resources').run_script('airtime-playout==1.0', 'airtime-playout')
Aug 15 18:41:07 debian11 airtime-playout[11261]:   File "/usr/local/lib/python3.9/dist-packages/pkg_resources/__init__.py", line 651, in run_script
Aug 15 18:41:07 debian11 airtime-playout[11261]:     self.require(requires)[0].run_script(script_name, ns)
Aug 15 18:41:07 debian11 airtime-playout[11261]:   File "/usr/local/lib/python3.9/dist-packages/pkg_resources/__init__.py", line 1448, in run_script
Aug 15 18:41:07 debian11 airtime-playout[11261]:     exec(code, namespace, namespace)
Aug 15 18:41:07 debian11 airtime-playout[11261]:   File "/usr/local/lib/python3.9/dist-packages/airtime_playout-1.0-py3.9.egg/EGG-INFO/scripts/airtime-playout", line 4, in <module>
Aug 15 18:41:07 debian11 airtime-playout[11261]:     runpy.run_module("pypo", run_name="__main__")
Aug 15 18:41:07 debian11 airtime-playout[11261]:   File "/usr/lib/python3.9/runpy.py", line 213, in run_module
Aug 15 18:41:07 debian11 airtime-playout[11261]:     return _run_code(code, {}, init_globals, run_name, mod_spec)
Aug 15 18:41:07 debian11 airtime-playout[11261]:   File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
Aug 15 18:41:07 debian11 airtime-playout[11261]:     exec(code, run_globals)
Aug 15 18:41:07 debian11 airtime-playout[11261]:   File "/usr/local/lib/python3.9/dist-packages/airtime_playout-1.0-py3.9.egg/pypo/__main__.py", line 247, in <module>
Aug 15 18:41:07 debian11 airtime-playout[11261]:     liquidsoap_startup_test()
Aug 15 18:41:07 debian11 airtime-playout[11261]:   File "/usr/local/lib/python3.9/dist-packages/airtime_playout-1.0-py3.9.egg/pypo/__main__.py", line 187, in liquidsoap_startup_test
Aug 15 18:41:07 debian11 airtime-playout[11261]:     liquidsoap_version_string = liquidsoap_get_info(
Aug 15 18:41:07 debian11 airtime-playout[11261]:   File "/usr/local/lib/python3.9/dist-packages/airtime_playout-1.0-py3.9.egg/pypo/timeout.py", line 40, in new_f
Aug 15 18:41:07 debian11 airtime-playout[11261]:     return __timeout(f, timeout, default, args, kwargs)
Aug 15 18:41:07 debian11 airtime-playout[11261]:   File "/usr/local/lib/python3.9/dist-packages/airtime_playout-1.0-py3.9.egg/pypo/timeout.py", line 24, in __timeout
Aug 15 18:41:07 debian11 airtime-playout[11261]:     if it.isAlive():
Aug 15 18:41:07 debian11 airtime-playout[11261]: AttributeError: 'InterruptableThread' object has no attribute 'isAlive'
Aug 15 18:41:07 debian11 systemd[1]: libretime-playout.service: Main process exited, code=exited, status=1/FAILURE
Aug 15 18:41:07 debian11 systemd[1]: libretime-playout.service: Failed with result 'exit-code'.
Aug 15 18:41:08 debian11 systemd[1]: libretime-playout.service: Scheduled restart job, restart counter is at 5.
Aug 15 18:41:08 debian11 systemd[1]: Stopped Libretime Playout Service.
Aug 15 18:41:08 debian11 systemd[1]: libretime-playout.service: Start request repeated too quickly.
Aug 15 18:41:08 debian11 systemd[1]: libretime-playout.service: Failed with result 'exit-code'.
Aug 15 18:41:08 debian11 systemd[1]: Failed to start Libretime Playout Service.

```

jooola commented:
It is simply a python compatibility problem, I'll push a fix.

paddatrapper commented:
Testing with vagrant - Bento does not ship a Debian 11 image yet - https://app.vagrantup.com/bento/

paddatrapper commented on .github/workflows/test.yml:
```diff
@@ -34,6 +34,12 @@ jobs:
             prerequisites-script: install-bionic.sh
 
+          # Similar to Debian Bullseye
+          - runs-on: ubuntu-20.04
```
Why add a 20.04 test and not a Debian 11 test? This will fail at the moment thanks to issues encountered in #1168

jooola replied:
Github Actions does not provide a Debian 11 platform to run on. So I tried to setup the most similar setup to what is bullseye.

I've been looking for a way to run all this test in docker images so we can pick ubuntu/debian/centos, but it requires some work, and it was part of the research in #1248.

paddatrapper replied:
Ah I see, yeah that does make it tricky

jooola commented:
In the CI, the installed python version is not picked up by our python test suite. Both jobs run with python 3.6, but one job is requesting python3.9. I guess I will not tweak too much in here, maybe #1248 is starting to become necessary.

jooola commented:
Just installed a fresh libretime on Debian bullseye, didn't have any problem. Everything is working as expected, I think this can be a candidat for this evening.

The only thing I am not sure about is some zend logs:

```
2021-08-17T14:45:54+02:00 ERR (3): station.radio.org [ErrorController.php:26 - errorAction()] - Cannot send headers; headers already sent in /usr/share/airtime/php/airtime_mvc/application/upgrade/Upgrades.php, line 243
2021-08-17T14:45:54+02:00 ERR (3): station.radio.org [ErrorController.php:27 - errorAction()] - #0 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Response/Abstract.php(115): Zend_Controller_Response_Abstract->canSendHeaders(true)
#1 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Action/Helper/ContextSwitch.php(283): Zend_Controller_Response_Abstract->setHeader('Content-Type', 'application/jso...')
#2 /usr/share/airtime/php/airtime_mvc/application/controllers/ApiController.php(80): Zend_Controller_Action_Helper_ContextSwitch->initContext()
#3 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Action.php(133): ApiController->init()
#4 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Dispatcher/Standard.php(281): Zend_Controller_Action->__construct(Object(Zend_Controller_Request_Http), Object(Zend_Controller_Response_Http), Array)
#5 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Front.php(954): Zend_Controller_Dispatcher_Standard->dispatch(Object(Zend_Controller_Request_Http), Object(Zend_Controller_Response_Http))
#6 /usr/share/airtime/php/vendor/zf1s/zend-application/library/Zend/Application/Bootstrap/Bootstrap.php(105): Zend_Controller_Front->dispatch()
#7 /usr/share/airtime/php/vendor/zf1s/zend-application/library/Zend/Application.php(384): Zend_Application_Bootstrap_Bootstrap->run()
#8 /usr/share/airtime/php/airtime_mvc/application/airtime-boot.php(84): Zend_Application->run()
#9 /usr/share/airtime/php/airtime_mvc/public/index.php(68): require_once('/usr/share/airt...')
#10 {main}
2021-08-17T14:45:54+02:00 ERR (3): station.radio.org [ErrorController.php:54 - errorAction()] - An internal application error has occurred.: Zend_Controller_Response_Exception: Cannot send headers; headers already sent in /usr/share/airtime/php/airtime_mvc/application/upgrade/Upgrades.php, line 243 in /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Response/Abstract.php:323
Stack trace:
#0 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Response/Abstract.php(115): Zend_Controller_Response_Abstract->canSendHeaders(true)
#1 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Action/Helper/ContextSwitch.php(283): Zend_Controller_Response_Abstract->setHeader('Content-Type', 'application/jso...')
#2 /usr/share/airtime/php/airtime_mvc/application/controllers/ApiController.php(80): Zend_Controller_Action_Helper_ContextSwitch->initContext()
#3 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Action.php(133): ApiController->init()
#4 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Dispatcher/Standard.php(281): Zend_Controller_Action->__construct(Object(Zend_Controller_Request_Http), Object(Zend_Controller_Response_Http), Array)
#5 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Front.php(954): Zend_Controller_Dispatcher_Standard->dispatch(Object(Zend_Controller_Request_Http), Object(Zend_Controller_Response_Http))
#6 /usr/share/airtime/php/vendor/zf1s/zend-application/library/Zend/Application/Bootstrap/Bootstrap.php(105): Zend_Controller_Front->dispatch()
#7 /usr/share/airtime/php/vendor/zf1s/zend-application/library/Zend/Application.php(384): Zend_Application_Bootstrap_Bootstrap->run()
#8 /usr/share/airtime/php/airtime_mvc/application/airtime-boot.php(84): Zend_Application->run()
#9 /usr/share/airtime/php/airtime_mvc/public/index.php(68): require_once('/usr/share/airt...')
#10 {main}
2021-08-17T14:45:54+02:00 WARN (4): Cannot modify header information - headers already sent by (output started at /usr/share/airtime/php/airtime_mvc/application/upgrade/Upgrades.php:243)
2021-08-17T14:45:54+02:00 ERR (3): station.radio.org [Logging.php:166 - loggingShutdownCallback()] - Uncaught ErrorException: Cannot modify header information - headers already sent by (output started at /usr/share/airtime/php/airtime_mvc/application/upgrade/Upgrades.php:243) in /usr/share/airtime/php/airtime_mvc/application/airtime-boot.php:93
Stack trace:
#0 /usr/share/airtime/php/vendor/zf1s/zend-log/library/Zend/Log.php(636): exception_error_handler(2, 'Cannot modify h...', '/usr/share/airt...', 93, Array)
#1 [internal function]: Zend_Log->errorHandler(2, 'Cannot modify h...', '/usr/share/airt...', 93, Array)
#2 /usr/share/airtime/php/airtime_mvc/application/airtime-boot.php(93): header('HTTP/1.1 500 In...', true, 500)
#3 /usr/share/airtime/php/airtime_mvc/public/index.php(68): require_once('/usr/share/airt...')
#4 {main}
  thrown at /usr/share/airtime/php/airtime_mvc/application/airtime-boot.php:93
array (
  'type' => 1,
  'message' => 'Uncaught ErrorException: Cannot modify header information - headers already sent by (output started at /usr/share/airtime/php/airtime_mvc/application/upgrade/Upgrades.php:243) in /usr/share/airtime/php/airtime_mvc/application/airtime-boot.php:93
Stack trace:
#0 /usr/share/airtime/php/vendor/zf1s/zend-log/library/Zend/Log.php(636): exception_error_handler(2, \'Cannot modify h...\', \'/usr/share/airt...\', 93, Array)
#1 [internal function]: Zend_Log->errorHandler(2, \'Cannot modify h...\', \'/usr/share/airt...\', 93, Array)
#2 /usr/share/airtime/php/airtime_mvc/application/airtime-boot.php(93): header(\'HTTP/1.1 500 In...\', true, 500)
#3 /usr/share/airtime/php/airtime_mvc/public/index.php(68): require_once(\'/usr/share/airt...\')
#4 {main}
  thrown',
  'file' => '/usr/share/airtime/php/airtime_mvc/application/airtime-boot.php',
  'line' => 93,
)
2021-08-17T14:45:54+02:00 ERR (3): station.radio.org [ErrorController.php:26 - errorAction()] - Cannot send headers; headers already sent in /usr/share/airtime/php/airtime_mvc/application/upgrade/Upgrades.php, line 243
2021-08-17T14:45:54+02:00 ERR (3): station.radio.org [ErrorController.php:27 - errorAction()] - #0 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Response/Abstract.php(115): Zend_Controller_Response_Abstract->canSendHeaders(true)
#1 /usr/share/airtime/php/vendor/zf1s/zend-view/library/Zend/View/Helper/Json.php(87): Zend_Controller_Response_Abstract->setHeader('Content-Type', 'application/jso...', true)
#2 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Action/Helper/Json.php(73): Zend_View_Helper_Json->json('{"airtime_versi...', false, true)
#3 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Action/Helper/Json.php(102): Zend_Controller_Action_Helper_Json->encodeJson(Array, false, true)
#4 /usr/share/airtime/php/airtime_mvc/application/controllers/ApiController.php(112): Zend_Controller_Action_Helper_Json->sendJson(Array)
#5 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Action.php(516): ApiController->versionAction()
#6 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Dispatcher/Standard.php(308): Zend_Controller_Action->dispatch('versionAction')
#7 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Front.php(954): Zend_Controller_Dispatcher_Standard->dispatch(Object(Zend_Controller_Request_Http), Object(Zend_Controller_Response_Http))
#8 /usr/share/airtime/php/vendor/zf1s/zend-application/library/Zend/Application/Bootstrap/Bootstrap.php(105): Zend_Controller_Front->dispatch()
#9 /usr/share/airtime/php/vendor/zf1s/zend-application/library/Zend/Application.php(384): Zend_Application_Bootstrap_Bootstrap->run()
#10 /usr/share/airtime/php/airtime_mvc/application/airtime-boot.php(84): Zend_Application->run()
#11 /usr/share/airtime/php/airtime_mvc/public/index.php(68): require_once('/usr/share/airt...')
#12 {main}
2021-08-17T14:45:54+02:00 ERR (3): station.radio.org [ErrorController.php:54 - errorAction()] - An internal application error has occurred.: Zend_Controller_Response_Exception: Cannot send headers; headers already sent in /usr/share/airtime/php/airtime_mvc/application/upgrade/Upgrades.php, line 243 in /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Response/Abstract.php:323
Stack trace:
#0 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Response/Abstract.php(115): Zend_Controller_Response_Abstract->canSendHeaders(true)
#1 /usr/share/airtime/php/vendor/zf1s/zend-view/library/Zend/View/Helper/Json.php(87): Zend_Controller_Response_Abstract->setHeader('Content-Type', 'application/jso...', true)
#2 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Action/Helper/Json.php(73): Zend_View_Helper_Json->json('{"airtime_versi...', false, true)
#3 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Action/Helper/Json.php(102): Zend_Controller_Action_Helper_Json->encodeJson(Array, false, true)
#4 /usr/share/airtime/php/airtime_mvc/application/controllers/ApiController.php(112): Zend_Controller_Action_Helper_Json->sendJson(Array)
#5 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Action.php(516): ApiController->versionAction()
#6 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Dispatcher/Standard.php(308): Zend_Controller_Action->dispatch('versionAction')
#7 /usr/share/airtime/php/vendor/zf1s/zend-controller/library/Zend/Controller/Front.php(954): Zend_Controller_Dispatcher_Standard->dispatch(Object(Zend_Controller_Request_Http), Object(Zend_Controller_Response_Http))
#8 /usr/share/airtime/php/vendor/zf1s/zend-application/library/Zend/Application/Bootstrap/Bootstrap.php(105): Zend_Controller_Front->dispatch()
#9 /usr/share/airtime/php/vendor/zf1s/zend-application/library/Zend/Application.php(384): Zend_Application_Bootstrap_Bootstrap->run()
#10 /usr/share/airtime/php/airtime_mvc/application/airtime-boot.php(84): Zend_Application->run()
#11 /usr/share/airtime/php/airtime_mvc/public/index.php(68): require_once('/usr/share/airt...')
#12 {main}
2021-08-17T14:45:54+02:00 WARN (4): Cannot modify header information - headers already sent by (output started at /usr/share/airtime/php/airtime_mvc/application/upgrade/Upgrades.php:243)
2021-08-17T14:45:54+02:00 ERR (3): station.radio.org [Logging.php:166 - loggingShutdownCallback()] - Uncaught ErrorException: Cannot modify header information - headers already sent by (output started at /usr/share/airtime/php/airtime_mvc/application/upgrade/Upgrades.php:243) in /usr/share/airtime/php/airtime_mvc/application/airtime-boot.php:93
Stack trace:
#0 /usr/share/airtime/php/vendor/zf1s/zend-log/library/Zend/Log.php(636): exception_error_handler(2, 'Cannot modify h...', '/usr/share/airt...', 93, Array)
#1 [internal function]: Zend_Log->errorHandler(2, 'Cannot modify h...', '/usr/share/airt...', 93, Array)
#2 /usr/share/airtime/php/airtime_mvc/application/airtime-boot.php(93): header('HTTP/1.1 500 In...', true, 500)
#3 /usr/share/airtime/php/airtime_mvc/public/index.php(68): require_once('/usr/share/airt...')
#4 {main}
  thrown at /usr/share/airtime/php/airtime_mvc/application/airtime-boot.php:93
array (
  'type' => 1,
  'message' => 'Uncaught ErrorException: Cannot modify header information - headers already sent by (output started at /usr/share/airtime/php/airtime_mvc/application/upgrade/Upgrades.php:243) in /usr/share/airtime/php/airtime_mvc/application/airtime-boot.php:93
Stack trace:
#0 /usr/share/airtime/php/vendor/zf1s/zend-log/library/Zend/Log.php(636): exception_error_handler(2, \'Cannot modify h...\', \'/usr/share/airt...\', 93, Array)
#1 [internal function]: Zend_Log->errorHandler(2, \'Cannot modify h...\', \'/usr/share/airt...\', 93, Array)
#2 /usr/share/airtime/php/airtime_mvc/application/airtime-boot.php(93): header(\'HTTP/1.1 500 In...\', true, 500)
#3 /usr/share/airtime/php/airtime_mvc/public/index.php(68): require_once(\'/usr/share/airt...\')
#4 {main}
  thrown',
  'file' => '/usr/share/airtime/php/airtime_mvc/application/airtime-boot.php',
  'line' => 93,
)
2021-08-17T14:45:57+02:00 INFO (6): station.radio.org You must explicitly start the session with Zend_Session::start() when session options are set to strict.
2021-08-17T14:45:59+02:00 INFO (6): station.radio.org Notifying RabbitMQ to send message to pypo
2021-08-17T14:45:59+02:00 ERR (3): station.radio.org [Logging.php:166 - loggingShutdownCallback()] - "continue" targeting switch is equivalent to "break". Did you mean to use "continue 2"? at /usr/share/airtime/php/vendor/zf1s/zend-form/library/Zend/Form.php:1690
array (
  'type' => 2,
  'message' => '"continue" targeting switch is equivalent to "break". Did you mean to use "continue 2"?',
  'file' => '/usr/share/airtime/php/vendor/zf1s/zend-form/library/Zend/Form.php',
  'line' => 1690,
)
2021-08-17T14:45:59+02:00 INFO (6): station.radio.org Registered Component: pypo@192.168.100.25
2021-08-17T14:45:59+02:00 INFO (6): station.radio.org Registered Component: show-recorder@192.168.100.25
2021-08-17T14:46:05+02:00 INFO (6): station.radio.org Database error: SQLSTATE[23505]: Unique violation: 7 ERROR:  duplicate key value violates unique constraint "cc_pref_subj_key_idx"
DETAIL:  Key (subjid, keystr)=(1, library_datatable) already exists.
```

Are the logs harmless ? Seems related to DB migration. This is the entire zendphp log file after the installation.

jooola commented:
Vagrant support is on the way https://github.com/chef/bento/pull/1367

paddatrapper commented:
We can use the official Debian vagrant images instead of waiting for the bento ones: https://app.vagrantup.com/debian/boxes/bullseye64

jooola commented:
What the heck happen here ?

paddatrapper commented:
Rebase caused issues? Try checkout https://github.com/LibreTime/libretime/commit/7a2e9102df3c03e37820b4cea5c1838533ca240c into a new version of tuis branch and rebase again?

jooola commented:
Weird, I get segfault with git reflog. Anyway, seems like I am not fresh today.

jooola commented:
I was wondering, what is the difference between the bento images and the official ones ? Do we really need to bento ones ?

paddatrapper commented:
I don't think we do. The only advantage I can see is that then we use the same provider for all our images. Using the official ones would be better I think

jooola commented:
I am not completly happy with the way we need N+1 commit/workflow run to make a change to the tests workflow.

As we don't have prebuild images of debian 11 we are stuck with a failing pipeline and no real way to run the tests.

paddatrapper commented:
> I  could open a PR to "prepare debian 11 support" and simply build the docker dev images. and use them in this PR. @paddatrapper But maybe you have an idea on how to improve this.

Hmm... Yeah probably a prepare PR is easiest. Short of building the docker image every time, I'm not sure anything else is possible

jooola commented:
Ref: This PR need #1334 to be able to run tests.

jooola commented:
So, unless @xabispacebiker has time to make the php libraries work with php7.4, this can be merged after some extra testing and a Vagrant test.

I'll install from this branch to a staging env see how it works out in the next weeks.

paddatrapper commented:
In vagrant I'm running into https://github.com/LibreTime/libretime/pull/1168#issuecomment-811723506, but I can't find the fix I mentioned...

jooola commented:
@paddatrapper What about using the CLI flags to check for playability and get the version ?

`liquidsoap --version`
`liquidsoap -r /path/to/file.mp3`

paddatrapper commented:
Fix for hanging on the version check - https://github.com/LibreTime/libretime/pull/1168/commits/ca0dbd0564cb86133a4df2d1fbb9fc2ede57fa35

jooola commented:
@paddatrapper I already implemented a way to get the version from `liquidsoap --version`, but I was wondering if your fix is better in order to maintain compatibility with older versions of liquidsoap ?

We should extract the version check in a function anyway to simplify the code.

I've been tweaking a lot of thing in order to make liquidsoap run on bullseye, this is why I opened #1338 as it seem to require more work than expected.

jooola commented:
Some recent changes:
- Reverted to use bento images again, they include the virtual box guest addition by default. I used libvirt but had some troubles.
- Removed the Sury repository since we will support php 7.4 anytime soon.
- Merged #1354 in this PR, so this PR depends on #1354 (I'll rebase and cleanup the history).

jooola commented:
I am running this PR in a staging environment as a testing platform for the editors of our radio station.

I'll post possible weird logs and push fixes as I encounter them. Nothing so far.

jooola commented:
@paddatrapper I would like to get this merged. I think we can review this.

The only thing I am not sure is the vagrant nfs4 option, it might break compatibility with older systems. How should we handle this ?

jooola commented:
Depends on #1424

jooola commented on .github/workflows/test.yml:
```diff
@@ -171,8 +178,14 @@ jobs:
+          - buster
+          - bullseye
+          - bionic
+          - focal
```
Focal was missing, is that right ?

paddatrapper replied:
Accidental. See #1430

paddatrapper commented:
> The only thing I am not sure is the vagrant nfs4 option, it might break compatibility with older systems. How should we handle this ?

I think we can assume development on a fairly modern platform and people developing with an older version of Vagrant can update their installation from another source or use multipass (where supported)

jooola commented:
I finally deployed the main branch to production on bullseye. We should have plenty of testing from now on.

Migration went smoothly, with db dump + restore, copied config files to /etc/airtime, and detached/attached the storage disk from old to new system.

crunchysteve commented:
So, why does any available LibreTime still refuse to install on bullseye? Am I going to have to completely retrograde my new linux rig?
